src/common/tusb_fifo.c

```diff
@@ -123,24 +270,91 @@ static inline void _ff_pull(tu_fifo_t* f, void * p_buffer, uint16_t rRel)
 }
 
 // get n items from FIFO WITHOUT updating read pointer
-static void _ff_pull_n(tu_fifo_t* f, void * p_buffer, uint16_t n, uint16_t rRel)
+static void _ff_pull_n(tu_fifo_t* f, void * p_buffer, uint16_t n, uint16_t rRel, tu_fifo_copy_mode_t copy_mode)
 {
-  if(rRel + n <= f->depth)       // Linear mode only
+  switch (copy_mode)
   {
-    memcpy(p_buffer, f->buffer + (rRel * f->item_size), n*f->item_size);
-  }
-  else      // Wrap around
-  {
-    uint16_t nLin = f->depth - rRel;
-
-    // Read data from linear part of buffer
-    memcpy(p_buffer, f->buffer + (rRel * f->item_size), nLin*f->item_size);
-
-    // Read data wrapped part
-    memcpy((uint8_t*)p_buffer + nLin*f->item_size, f->buffer, (n - nLin) * f->item_size);
+    case TU_FIFO_COPY_INC:
+      if(n <= f->depth-rRel)
+      {
+        // Linear mode only
+        memcpy(p_buffer, f->buffer + (rRel * f->item_size), n*f->item_size);
+      }
+      else
+      {
+        // Wrap around
+        uint16_t nLin = f->depth - rRel;
+
+        // Read data from linear part of buffer
+        memcpy(p_buffer, f->buffer + (rRel * f->item_size), nLin*f->item_size);
+
+        // Read data wrapped part
+        memcpy((uint8_t*)p_buffer + nLin*f->item_size, f->buffer, (n - nLin) * f->item_size);
+      }
+      break;
+
+    case TU_FIFO_COPY_CST_FULL_WORDS:
+
+      if(n <= f->depth-rRel)
+      {
+        // Linear mode only
+        _tu_fifo_write_to_const_dst_ptr_in_full_words(p_buffer, f->buffer + (rRel * f->item_size), n*f->item_size);
```

Comment:
_tu_fifo_write() is rather misleading, should be read to const_dst, although it is already write to that address, this is called within the tu_fifo_read(), should be named in consistent with the data direction from/to fifo. I will do the rename.

Reply:
Jep, true... this comes from coding in the middle of the night ^^. It should be `read_to_const()` and `write_from_const()`

Reply:
@PanRe no worries, I am renaming as well as applying the new tu_unaligned_write32/read32(). Seem to work well with stm32f4 (ofc since it is M4).